fv3core/stencils/d_sw.py

```diff
@@ -93,76 +94,225 @@ def horizontal_relative_vorticity_from_winds(
-        ke[0, 0, 0] = 0.5 * (ke + ub * vb)
+    return vt, ut, vorticity
+
+
+@gtscript.function
+def add_dw(w, dw, damp_w):
+    w = w + dw if damp_w > 1e-5 else w
+    return w
+
+
+@gtscript.function
+def ke_from_bwind(ke, ub, vb):
+    return 0.5 * (ke + ub * vb)
+
+
+@gtscript.function
+def adjust_w_and_qcon(w, delp, dw, damp_w, q_con):
+    from __externals__ import local_ie, local_is, local_je, local_js
+
+    with horizontal(region[local_is : local_ie + 1, local_js : local_je + 1]):
+        w = w / delp
+        w = add_dw(w, dw, damp_w)
+        q_con = q_con / delp
+
+    return w, q_con
+
+
+@gtscript.function
+def all_corners_ke(ke, u, v, ut, vt, dt):
+    from __externals__ import i_end, i_start, j_end, j_start
+
+    # Assumption: not __INLINED(spec.grid.nested)
+    with horizontal(region[i_start, j_start]):
+        ke = corners.corner_ke(ke, u, v, ut, vt, dt, 0, 0, -1, 1)
+    with horizontal(region[i_end + 1, j_start]):
+        ke = corners.corner_ke(ke, u, v, ut, vt, dt, -1, 0, 0, -1)
+    with horizontal(region[i_end + 1, j_end + 1]):
+        ke = corners.corner_ke(ke, u, v, ut, vt, dt, -1, -1, 0, 1)
+    with horizontal(region[i_start, j_end + 1]):
+        ke = corners.corner_ke(ke, u, v, ut, vt, dt, 0, -1, -1, -1)
+
+    return ke
 
 
 @gtstencil()
-def ub_from_vort(vort: sd, ub: sd):
+def ke_horizontal_vorticity_w_qcon_adjust(
+    ke: FloatField,
+    u: FloatField,
+    v: FloatField,
+    ub: FloatField,
+    vb: FloatField,
+    ut: FloatField,
+    vt: FloatField,
+    dx: FloatField,
+    dy: FloatField,
+    rarea: FloatField,
+    vorticity: FloatField,
+    w: FloatField,
+    delp: FloatField,
+    dw: FloatField,
+    q_con: FloatField,
+    dt: float,
+    damp_w: float,
+):
+
     with computation(PARALLEL), interval(...):
-        ub[0, 0, 0] = vort - vort[1, 0, 0]
+        ke = ke_from_bwind(ke, ub, vb)
+        ke = all_corners_ke(ke, u, v, ut, vt, dt)
+        vt, ut, vorticity = horizontal_relative_vorticity_from_winds_f(
+            u, v, ut, vt, dx, dy, rarea, vorticity
+        )
+        w, q_con = adjust_w_and_qcon(w, delp, dw, damp_w, q_con)
 
 
 @gtstencil()
-def vb_from_vort(vort: sd, vb: sd):
+def not_inlineq_pressure_and_vbke(
+    gx: FloatField,
+    gy: FloatField,
+    rarea: FloatField,
+    fx: FloatField,
+    fy: FloatField,
+    pt: FloatField,
+    delp: FloatField,
+    vc: FloatField,
+    uc: FloatField,
+    cosa: FloatField,
+    rsina: FloatField,
+    vt: FloatField,
+    vb: FloatField,
+    dt4: float,
+    dt5: float,
+):
+    from __externals__ import (
+        i_end,
+        i_start,
+        j_end,
+        j_start,
+        local_ie,
+        local_is,
+        local_je,
+        local_js,
+    )
+
     with computation(PARALLEL), interval(...):
-        vb[0, 0, 0] = vort - vort[0, 1, 0]
+        if __INLINED(spec.namelist.inline_q == 0):
+            with horizontal(region[local_is : local_ie + 1, local_js : local_je + 1]):
```

Comment:
Nice work! Would you mind adding a "TODO: only needed for d_sw validation"
 -- it looks like not having this, DynCore would still work, and it will help us in a subsequent pass to know if some regions can be removed simply by restricting the validation region.

Reply:
Done!